### PR TITLE
add snapshots to savegames

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Targeted post next bugfix release!

**Comment by fabiangreffrath** [comment]
To test:
- Do snapshots show for 16:10 aspect retios in widescreen mode (they should)?
- Are snapshots really limited to 320x200 px (they should) or am I missing some `>> hires` bit-shifting anywhere during snapshot window dimensions calculation?

**Comment by rfomin** [comment]
Maybe remove the skull and somehow cram it into 4:3? I like this (screenshot from ACE Engine):
image

**Comment by fabiangreffrath** [comment]
The description fields are supposed to be 200 px long, else the full description won't fit. I know this screenshot, but I wouldn't want to cut these fields shorter than they used to be.

**Comment by rfomin** [comment]
Okay, but I still think we should remove the skull. Perhaps replace it with `>`.

**Comment by fabiangreffrath** [comment]
He, what, why? The skull is Vanilla Doom and many PWADs replace it with something else for a consistent style.

**Comment by rfomin** [comment]
Because it's too big. Screenshot from GZDoom:
image

**Comment by rfomin** [comment]
> Because it's too big. Screenshot from GZDoom: image

I mean remove it only in Save/Load menus.

File: src/m_menu.c [hunk 1]
```diff
@@ -797,12 +800,72 @@ menu_t LoadDef =
...
+  const int snapshot_x = MAX((WIDESCREENDELTA + SaveDef.x + SKULLXOFF - snapshot_width) / 2, 8);
+  const int snapshot_y = LoadDef.y + MAX((load_end * LINEHEIGHT - snapshot_height) * n / load_end, 0);
+
+  // [FG] a snapshot window smaller than 80*48 px is considered too small
+  if (snapshot_width < ORIGWIDTH/4)
+    return;
+
+  if (!M_DrawSnapshot(n, snapshot_x, snapshot_y, snapshot_width, snapshot_height))
+  {
+    char *na = "n/a";
```
Comment (fabiangreffrath) [inline src/m_menu.c:828]: After this is merged, I'd like to work on const-correctness for `M_WriteText()`, `M_StringWidth()`, `M_StringHeight()` and the like which still get passed non-const `char *` as arguments.

**Comment by fabiangreffrath** [comment]
I like the date/time part on the GZDOom screenshot and consider adding it below the snapshots. This, or the leveltime information. Not sure what's more important, but we certainly don't have enough space for both (we are bound to the metrics of the HUD fonts).

**Comment by rfomin** [comment]
> I like the date/time part on the GZDOom screenshot and consider adding it below the snapshots. This, or the leveltime information

I think date/time is more useful.

To be honest, I like the implementation of the ACE Engine the most - almost vanilla look and bigger snapshots.

**Comment by fabiangreffrath** [comment]
> I think date/time is more useful.

Alright, thanks!

> To be honest, I like the implementation of the ACE Engine the most - almost vanilla look and bigger snapshots.

Sure, but the downside is that the descriptions lines have to get shrinked from 200px to <160 px.

File: src/m_menu.c [hunk 2]
```diff
@@ -797,12 +800,78 @@ menu_t LoadDef =
...
+                snapshot_y + snapshot_height/2 - M_StringHeight(txt)/2,
+                txt);
+  }
+
+  txt = M_GetSavegameTime(n);
+  M_DrawString(snapshot_x + snapshot_width/2 - M_GetPixelWidth(txt)/2 - WIDESCREENDELTA,
+               snapshot_y + snapshot_height + M_StringHeight(txt),
+               CR_GOLD, txt);
+
+  // [FG] draw the view border around the snapshot
```
Comment (rfomin) [inline src/m_menu.c:840]: I see that this code is taken from `R_FillBackScreen()`. How about a new function (eg `R_DrawBorder()`) to reduce code duplication?

**Review by rfomin — COMMENTED** [review]
One little nitpick - most other ports don't draw STBAR/ui in snapshots. Have you thought about it?

File: src/m_snapshot.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    return false;
+
+  return true;
+}
+
+void M_ReadSavegameTime (int i, char *name)
+{
+  struct stat st;
+
+  if (stat(name, &st) == -1)
```
Comment (rfomin) [inline src/m_snapshot.c:83]: Please use `M_stat()` from `m_io.h`.

Comment (fabiangreffrath) [reply src/m_snapshot.c:83]: Sure, will do!

File: src/m_menu.c (hunk 2 above)
Comment (fabiangreffrath) [reply src/m_menu.c:840]: > One little nitpick - most other ports don't draw STBAR/ui in snapshots. Have you thought about it?

No, I haven't thought about it.

> I see that this code is taken from `R_FillBackScreen()`. How about a new function (eg `R_DrawBorder()`) to reduce code duplication?

Very good idea, will work on it post-PR.

**Pull request merged**